File: plugins-sdk/src/api/storage.ts
import { invoke } from '../core/ipc';
import { dispatch } from '../core/dispatch';

export interface StorageOptions {
  // 预留配置选项
}

export interface StorageError extends Error {
  name: 'StorageError';
  key?: string;
}

// 错误工厂函数
export function createStorageError(message: string, key?: string): StorageError {
  const error = new Error(message) as StorageError;
  error.name = 'StorageError';
  error.key = key;
  return error;
}

// 类型检查函数
export function isStorageError(error: any): error is StorageError {
  return error && error.name === 'StorageError';
}

/**
 * 设置一个键值对到存储中
 * @param key 存储键
 * @param value 存储值（会自动序列化）
 */
export function setItem(key: string, value: any): Promise<void> {
  return dispatch({
    webview: () => invoke("plugin_storage_set", { key, value }),
    headless: () => invoke("plugin_storage_set", { key, value }),
  });
}

/**
 * 从存储中获取一个值
 * @param key 存储键
 * @returns 存储的值，如果不存在则返回 null
 */
export function getItem<T = any>(key: string): Promise<T | null> {
  return dispatch({
    webview: () => invoke<T | null>("plugin_storage_get", { key }),
    headless: () => invoke<T | null>("plugin_storage_get", { key }),
  });
}

/**
 * 从存储中删除一个键
 * @param key 要删除的键
 */
export function removeItem(key: string): Promise<void> {
  return dispatch({
    webview: () => invoke("plugin_storage_remove", { key }),
    headless: () => invoke("plugin_storage_remove", { key }),
  });
}

/**
 * 清空当前插件的所有存储数据
 */
export function clear(): Promise<void> {
  return dispatch({
    webview: () => invoke("plugin_storage_clear"),
    headless: () => invoke("plugin_storage_clear"),
  });
}

/**
 * 获取当前插件存储的所有键
 * @returns 所有键的数组
 */
export function keys(): Promise<string[]> {
  return dispatch({
    webview: () => invoke<string[]>("plugin_storage_keys"),
    headless: () => invoke<string[]>("plugin_storage_keys"),
  });
}

/**
 * 批量设置多个键值对
 * @param items 要设置的键值对对象
 */
export function setItems(items: Record<string, any>): Promise<void> {
  return dispatch({
    webview: () => invoke("plugin_storage_set_items", { items }),
    headless: () => invoke("plugin_storage_set_items", { items }),
  });
}

/**
 * 批量获取多个键的值
 * @param keys 要获取的键数组
 * @returns 包含键值对的对象
 */
export function getItems<T = any>(keys: string[]): Promise<Record<string, T>> {
  return dispatch({
    webview: () => invoke<Record<string, T>>("plugin_storage_get_items", { keys }),
    headless: () => invoke<Record<string, T>>("plugin_storage_get_items", { keys }),
  });
}

// 导出默认的存储对象
export const storage = {
  setItem,
  getItem,
  removeItem,
  clear,
  keys,
  setItems,
  getItems
};

File: src-tauri/src/plugin_api/storage.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::AppHandle;
use tauri_plugin_store::{Error, StoreBuilder};

// 线程本地存储用于保存当前插件ID
thread_local! {
    static CURRENT_PLUGIN_ID: std::cell::RefCell<Option<String>> = const { std::cell::RefCell::new(None) };
}

// 设置当前插件ID（在插件调用开始时调用）
pub fn set_current_plugin_id(plugin_id: String) {
    CURRENT_PLUGIN_ID.with(|id| {
        *id.borrow_mut() = Some(plugin_id);
    });
}

// 清除当前插件ID（在插件调用结束时调用）
pub fn clear_current_plugin_id() {
    CURRENT_PLUGIN_ID.with(|id| {
        *id.borrow_mut() = None;
    });
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StorageError {
    pub name: String,
    pub message: String,
}

impl From<String> for StorageError {
    fn from(message: String) -> Self {
        StorageError {
            name: "StorageError".to_string(),
            message,
        }
    }
}

impl From<&str> for StorageError {
    fn from(message: &str) -> Self {
        StorageError {
            name: "StorageError".to_string(),
            message: message.to_string(),
        }
    }
}

impl From<Error> for StorageError {
    fn from(error: Error) -> Self {
        StorageError {
            name: "StorageError".to_string(),
            message: error.to_string(),
        }
    }
}

// 获取插件存储路径
fn get_plugin_store_path(plugin_id: &str) -> String {
    format!("plugin_data/{}/storage.json", plugin_id)
}

// 获取当前执行插件的 ID
fn get_current_plugin_id(_app: &AppHandle) -> Result<String, StorageError> {
    CURRENT_PLUGIN_ID.with(|id| {
        id.borrow().clone().ok_or_else(|| {
            StorageError::from("No plugin context found. Storage API must be called from within a plugin execution context.")
        })
    })
}

#[tauri::command]
pub async fn plugin_storage_set(
    app: AppHandle,
    key: String,
    value: serde_json::Value,
) -> Result<(), StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;

    store.set(key.clone(), value);
    store.save()?;

    println!("[Storage] Set key '{}' for plugin '{}'", key, plugin_id);
    Ok(())
}

#[tauri::command]
pub async fn plugin_storage_get(
    app: AppHandle,
    key: String,
) -> Result<Option<serde_json::Value>, StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;

    store.reload()?;
    let value = store.get(&key).map(|v| v.clone());

    println!(
        "[Storage] Get key '{}' for plugin '{}': {:?}",
        key,
        plugin_id,
        value.is_some()
    );
    Ok(value)
}

#[tauri::command]
pub async fn plugin_storage_remove(
    app: AppHandle,
    key: String,
) -> Result<(), StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;

    store.delete(&key);
    store.save()?;

    println!("[Storage] Removed key '{}' for plugin '{}'", key, plugin_id);
    Ok(())
}

#[tauri::command]
pub async fn plugin_storage_clear(
    app: AppHandle,
) -> Result<(), StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;

    store.clear();
    store.save()?;

    println!("[Storage] Cleared all data for plugin '{}'", plugin_id);
    Ok(())
}

#[tauri::command]
pub async fn plugin_storage_keys(
    app: AppHandle,
) -> Result<Vec<String>, StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;

    store.reload()?;
    let keys: Vec<String> = store.keys().iter().cloned().collect();

    println!("[Storage] Got {} keys for plugin '{}'", keys.len(), plugin_id);
    Ok(keys)
}

#[tauri::command]
pub async fn plugin_storage_set_items(
    app: AppHandle,
    items: HashMap<String, serde_json::Value>,
) -> Result<(), StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;
    let items_len = items.len();

    for (key, value) in items {
        store.set(key, value);
    }
    store.save()?;

    println!(
        "[Storage] Set {} items for plugin '{}'",
        items_len, plugin_id
    );
    Ok(())
}

#[tauri::command]
pub async fn plugin_storage_get_items(
    app: AppHandle,
    keys: Vec<String>,
) -> Result<HashMap<String, serde_json::Value>, StorageError> {
    let plugin_id = get_current_plugin_id(&app)?;
    let store_path = get_plugin_store_path(&plugin_id);
    let store = StoreBuilder::new(&app, store_path).build()?;

    store.reload()?;
    let mut result = HashMap::new();
    for key in keys {
        if let Some(value) = store.get(&key) {
            result.insert(key, value.clone());
        }
    }

    println!(
        "[Storage] Got {} items for plugin '{}'",
        result.len(),
        plugin_id
    );
    Ok(result)
}